### ISSUE TITLE
[New feature] Ability to restrict connection between containers

### ISSUE THREAD

**Issue by Yivan** [moby/moby#37309]
Use case:
We are using Docker to create a specific environement for each of our project.
This means each project containers should not be linked with other projet containers.
Each project for instance as an apache container, a php container, a mysql container, elasticsearch container, etc.
We wanted each php container access only to the mysql container and elasticsearch container of the project for instance.

I spend lot of time and tries to find a way to resolve this classic case with proper, simple and elegant solution... with no luck. So i think there is something missing and docker could be improved maybe.

Actual solutions:
- create a network for each project so containers of the projet can communicate only between them. This approach is limited as (without swarm which has overlay network and can create 65k network)
  - number of network is limited to 31 (for reference: https://loomchild.net/2016/09/04/docker-can-create-only-31-networks-on-a-single-machine/), so no more than 31 projects possible.
  - create many networks tends to make the host slower (if you make a loop to create 31 network, the first ones will create fast, but the last ones it will be long ; ))

- use the default network bridge -icc option and link container:
  - using "link" is deprecated (https://docs.docker.com/network/links/)
  - -icc seems to give some problem on some hosts (need br_netfilter activated on the host...)
  - this solution only work with the default bridge network, not custom create ones

- create a custom network bridge with -icc set to false:
  - "link" deprecated feature don't work for custom network with icc false
  - need to add manually rule in iptables (not possible with Docker for Windows for instance as we cannot accesss to the VM without some really dirty hackink involving nscenter and privilege excalation)

- use a reverse-proxy (traefik, nginx reverse proxy, ...):
  - work ok for http connection (or protocol supported by the reverse proxy), but tiedous to configure as for instance with traefik you need to limit by IP and in docker-compose you don't kown the ip...
  - we have to handle service url, etc...

**If you think of other solutions you can propose them in this issue please so this could help people. Thanks.**

Proposal:
An easy way to restrict connection on a container by only some container.
First we could create a custom network disabling interconnectivity:
`docker network create custom_global_network -o "com.docker.network.bridge.enable_icc"="false"`

Inside a docker-compose file we could have something like a new option `network_container_allow` (maybe could be an option of the network bridge):

```
version: "3.5"

services:
    php:
        ...
        networks:
            - custom_global_network

    mysql:
        ...
        networks:
            - custom_global_network:
              - allow: "php"

...
```

This way the 31 network limitation is no more a problem (we can create as many docker project as needed), just to need to create one network (light for the host).
The option `allow` will automatically configure iptable so connection beetween containers will be possible. As they are already on the same network they can ping directly from the container name.

**Comment by Yivan** [comment]
Just to give another soltuion working for MySQL: using socket on shared volume work great, not network to configure, you can leave MySQL container with no network at all.
But this dont work for elasticsearch has it doesn't support sucket, but just http connection.
So this is not a generic solution but can be applied on service supporting socket (that is what we use actually for mysql, redis and php, all three using shared volume per project with the sockets file inside).

**Comment by Yivan** [comment]
Thanks @cpuguy83 for you answer.

Yes you are totally right, and like in the link in my original post, you can create manually more than 31:
`for net in {1..50}; do docker network create -d bridge –subnet=172.18.${net}.0/24 net${net} ; done`

But in this case you need to handle manually network creation (can be easily adressed), but network subnet availabilities too (available network subnet for creation) for each project and this last one is the tricky part (when you up a docker compose stack, you will need to look for available subnet manually...).
Docker already assign automatically this by just running `docker network create -d bridge mynetwork`, but is limited to 31 in this case.
**Could this limit be improved ?** (something like 1000 network so it will be enough for all case... more than 1000 network, the host will have performance pbl i think ; )).
But actually 31 networks is really short... I really like the designed network features, but this limitation is for some use-case a blocking point.

**Comment by cpuguy83** [comment]
You can specify network pools/sizes to use as a daemon option so "docker network create" doesn't need to unless they really want control.

**Comment by cpuguy83** [comment]
I'm guessing this will only be included in 18.06.
Implementation was merged here: https://github.com/moby/moby/pull/36396

**Comment by thaJeztah** [comment]
Let me close this ticket for now, as it looks like it went stale.
